File: src/main/java/corp/soyummy/subscribes/SubscribeController.java
package corp.soyummy.subscribes;

import org.springframework.web.bind.annotation.*;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;

import lombok.RequiredArgsConstructor;
import jakarta.validation.Valid;
import java.util.List;

import corp.soyummy.subscribes.dto.SubscribeCreateDto;
import corp.soyummy.subscribes.dto.SubscribeUpdateDto;
import corp.soyummy.auth.User;
import corp.soyummy.errors.ResourceNotFoundException;
import static corp.soyummy.constants.Constants.VERSION;

@RestController
@RequestMapping(VERSION + "subscribes")
@RequiredArgsConstructor
public class SubscribeController {

    private final SubscribeService subscribeService;

    // GET http://localhost:8080/api/v1/subscribes/
    @GetMapping
    public ResponseEntity<List<Subscribe>> getAllSubscribes() {
        List<Subscribe> subscribes = subscribeService.getAllSubscribes();
        return ResponseEntity.ok(subscribes);
    }

    // GET http://localhost:8080/api/v1/subscribes/{id}
    @GetMapping("/{id}")
    public ResponseEntity<Subscribe> getSubscribeById(@PathVariable String id) {
        Subscribe subscribe = subscribeService.getSubscribeById(id);
        return ResponseEntity.ok(subscribe);
    }

    // GET http://localhost:8080/api/v1/subscribes/search?email=anna
    @GetMapping("/search")
    public ResponseEntity<List<Subscribe>> findByEmail(@RequestParam String email) {
        List<Subscribe> subscribes = subscribeService.findByEmail(email);
        return ResponseEntity.ok(subscribes);
    }

    // GET http://localhost:8080/api/v1/subscribes/isSubscribed?userId={userId}&email={email}
    @GetMapping("/isSubscribed")
    public ResponseEntity<Boolean> isSubscribed(
            @RequestParam String userId,
            @RequestParam String email) {
        boolean isSubscribed = subscribeService.isSubscribed(userId, email);
        return ResponseEntity.ok(isSubscribed);
    }

    // POST http://localhost:8080/api/v1/subscribes
    @PostMapping
    public ResponseEntity<Subscribe> createSubscribe(
            @Valid @RequestBody SubscribeCreateDto subscribeCreateDto,
            @AuthenticationPrincipal User currentUser) {
        subscribeCreateDto.setOwner(currentUser.getId());
        Subscribe subscribe = subscribeService.createSubscribe(subscribeCreateDto);
        return ResponseEntity.ok(subscribe);
    }

    // PUT http://localhost:8080/api/v1/subscribes/{id}
    @PutMapping("/{id}")
    public ResponseEntity<Subscribe> updateSubscribe(
            @PathVariable String id,
            @Valid @RequestBody SubscribeUpdateDto subscribeUpdateDto,
            @AuthenticationPrincipal User currentUser) {
        Subscribe existingSubscribe = subscribeService.getSubscribeById(id);
        if (!existingSubscribe.getOwnerId().equals(currentUser.getId())) {
            throw new ResourceNotFoundException("Subscription not found for current user");
        }

        Subscribe subscribe = subscribeService.updateSubscribe(id, subscribeUpdateDto);
        return ResponseEntity.ok(subscribe);
    }

    // POST http://localhost:8080/api/v1/subscribes/unsubscribe?userId={userId}&email={email}
    @PostMapping("/unsubscribe")
    public ResponseEntity<Void> unsubscribe(
            @RequestParam String userId,
            @RequestParam String email,
            @AuthenticationPrincipal User currentUser) {
        if (!subscribeService.isSubscribed(currentUser.getId(), email)) {
            throw new ResourceNotFoundException("Subscription not found for current user");
        }

        subscribeService.unsubscribe(userId, email);
        return ResponseEntity.noContent().build();
    }

    // DELETE http://localhost:8080/api/v1/subscribes/{id}
    @DeleteMapping("/{id}")
    public ResponseEntity<Void> deleteSubscribe(@PathVariable String id) {
        subscribeService.deleteSubscribe(id);
        return ResponseEntity.ok().build();
    }
}

File: src/main/kotlin/com/soyummy/subscribes/SubscribeController.kt
package com.soyummy.subscribes

import org.springframework.web.bind.annotation.*
import org.springframework.security.access.annotation.Secured
import org.springframework.http.ResponseEntity
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Sort
import com.soyummy.common.dto.PageResponse
import jakarta.validation.Valid

import com.soyummy.subscribes.dto.SubscribeCreateDto
import com.soyummy.subscribes.dto.SubscribeUpdateDto
import com.soyummy.auth.User
import com.soyummy.constants.Constants.VERSION

@RestController
@RequestMapping("${VERSION}/subscribes")
class SubscribeController(private val subscribeService: SubscribeService) {

    // GET http://localhost:8080/api/v1/subscribes?page=3&size=4
    @Secured("ROLE_ADMIN")
    @GetMapping
    fun getAllSubscribes(
        @RequestParam(defaultValue = "0") page: Int,
        @RequestParam(defaultValue = "10") size: Int,
        @RequestParam(defaultValue = "email,asc") sort: String
    ): ResponseEntity<PageResponse<Subscribe>> {
        val sortParams = sort.split(",")
        val sortDirection = if (sortParams.size > 1 && sortParams[1].lowercase() == "desc") {
            Sort.Direction.DESC
        } else {
            Sort.Direction.ASC
        }
        val sortField = sortParams[0]

        val pageable = PageRequest.of(page, size, Sort.by(sortDirection, sortField))
        val result = subscribeService.getAllSubscribes(pageable)

        return ResponseEntity.ok(result)
    }

    // GET http://localhost:8080/api/v1/subscribes/search?email=anna
    @Secured("ROLE_ADMIN")
    @GetMapping("/search")
    fun findByEmail(@RequestParam email: String): ResponseEntity<List<Subscribe>> =
        ResponseEntity.ok(subscribeService.findByEmail(email))

    // GET http://localhost:8080/api/v1/subscribes/{id}
    @GetMapping("/{id}")
    fun getSubscribeById(
        @PathVariable id: String,
        @AuthenticationPrincipal currentUser: User?
    ): ResponseEntity<Subscribe> =
        ResponseEntity.ok(subscribeService.getSubscribeById(id, currentUser))

    // GET http://localhost:8080/api/v1/subscribes/isSubscribed?userId={userId}&email={email}
    @GetMapping("/isSubscribed")
    fun isSubscribed(
        @RequestParam userId: String,
        @RequestParam email: String,
        @AuthenticationPrincipal currentUser: User?
    ): ResponseEntity<Boolean> =
        ResponseEntity.ok(subscribeService.isSubscribed(userId, email, currentUser))

    // POST http://localhost:8080/api/v1/subscribes
    @PostMapping
    fun createSubscribe(
        @Valid @RequestBody subscribeCreateDto: SubscribeCreateDto,
        @AuthenticationPrincipal currentUser: User?
    ): ResponseEntity<Subscribe> =
        ResponseEntity.ok(subscribeService.createSubscribe(subscribeCreateDto, currentUser))

    // PUT http://localhost:8080/api/v1/subscribes/{id}
    @PutMapping("/{id}")
    fun updateSubscribe(
        @PathVariable id: String,
        @Valid @RequestBody subscribeUpdateDto: SubscribeUpdateDto,
        @AuthenticationPrincipal currentUser: User?
    ): ResponseEntity<Subscribe> =
        ResponseEntity.ok(subscribeService.updateSubscribe(id, subscribeUpdateDto, currentUser))

    // POST http://localhost:8080/api/v1/subscribes/unsubscribe?userId={userId}&email={email}
    @PostMapping("/unsubscribe")
    fun unsubscribe(
        @RequestParam userId: String,
        @RequestParam email: String,
        @AuthenticationPrincipal currentUser: User?
    ): ResponseEntity<Void> {
        subscribeService.unsubscribe(userId, email, currentUser)
        return ResponseEntity.noContent().build()
    }

    // DELETE http://localhost:8080/api/v1/subscribes/{id}
    @DeleteMapping("/{id}")
    fun deleteSubscribe(
        @PathVariable id: String,
        @AuthenticationPrincipal currentUser: User?
    ): ResponseEntity<Void> {
        subscribeService.deleteSubscribe(id, currentUser)
        return ResponseEntity.ok().build()
    }
}